basic conversion to C++20

neheb opened the pull request.

neheb committed:
easyaccess const changes

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
change default C++ to 20

Now that older platforms are going away, we can migrate.

Also get rid of std::filesystem handling. Not needed with newer
compilers.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
get rid of startsWith

Signed-off-by: Rosen Penev <rosenp@gmail.com>

mandree commented:
> Hrm I’m unsure of this PR. Compiler minimums have to increase for everything.

For library code it is usually good to use newer standards only in the implementation, but not in the interface. However I do not know how to map this in Windows or whenever something links to features you need a newer standard C++/template library because then people still need to upgrade.  So the safe way would be to make the library stuff compatible from whatever minimum C++ standard version is desired (11/14/17) but pushing all users to C++20 might make exiv2 unsuitable for many of those who use it today.

neheb commented:
Right. The headers are still kept at C++11. On that note, if constexpr should be removed from them as compilers are not free to add it implicitly as I originally thought, making the templates much less efficient.

As far as Windows is concerned, that's not an issue. In the case of MSVC, the libraries are upgrade-able, I actually don't know if older MSVC can even compile exiv2. I remember MSVC2019 or something just crashing when trying to compile exiv2. No errors printed.

On that note, there's currently an issue where Exiv2 is incompatible with msvcrt. Given that it's deprecated by Microsoft, I see no value in supporting it. See https://github.com/Exiv2/exiv2/issues/2637

neheb commented:
This is still draft after all.

neheb commented:
Yeah I'm not sure of this PR anymore. People are still using GCC7 to compile Exiv2. std::endian comes with GCC8 and starts_with GCC9. Let's not mention other projects that are on C++14 to allow building with CentOS7. This is going to be a draft for a while....

neheb commented on src/easyaccess.cpp:
```diff
@@ -16,6 +16,16 @@
+template <size_t N, const char* const (&keys)[N]>
+ExifData::const_iterator findMetadatum(const ExifData& ed) {
+  for (const auto& k : keys) {
+    auto pos = ed.findKey(ExifKey(k));
```
wrong.

mandree commented:
Kindly do not pay too much attention on the decrepit Ubuntu LTS versions, that LTS promise holds only for "main"-series packages but not for "universe" what the average desktop is constituted from, so just ignore old Ubuntu.  No upstream projects has to cater for museum software on the desktops... is software as Exiv2 normally (unless on a rolling release) really rebuilt frequently by end users? 

Or will the new software only propagate into distros briefly before time their release slush/feature freeze?

neheb commented:
Rebased. I'm still not sold on killing support for GCC8.

neheb commented on .github/workflows/on_PR_meson.yaml:
```diff
@@ -23,15 +23,15 @@ jobs:
           meson compile -C "${{github.workspace}}/build"
           meson test -C "${{github.workspace}}/build"
   Ubuntu-clang:
     runs-on: ubuntu-20.04
```
Hrm that’s unfortunate.

neheb committed:
meson CI: update to 22.04

20.04 is going away in February. Update it in advance. Requires changes
to minimum compilers.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented on cmake/compilerFlags.cmake:
```diff
@@ -1,9 +1,7 @@
-endif()
+set(CMAKE_CXX_STANDARD 20)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 set(CMAKE_CXX_EXTENSIONS ON)
```
I remember now. It was some C function in exiv2.

neheb replied:
Correct

neheb commented:
Hrm that's harder to do looks like. fp_ needs complete removal.

mandree commented:
Avoiding strnlen where suitable (with stringviews or whatnot as input where input strings aren't C strings) is probably easiest.

See https://www.cygwin.com/cygwin-api/compatibility.html - this claims strnlen to be in POSIX.
See https://gitlab.kitware.com/cmake/cmake/-/issues/20454 - cmake still discusses what do to about extensions.

But: Did we also just **also** disable POSIX extensions to the language when disabling the gnu extensions? That looks similar, but is not the same. 

Re-enabling those could happen through either of (one is sufficient!) - before the first include, or through CPPFLAGS/CFLAGS/CXXFLAGS:

`#define _XOPEN_SOURCE 700`  
or
`#define _POSIX_C_SOURCE 200809L`

neheb committed:
remove std::experimental::filesystem

This breaks support for older compilers, especially ones that don't have
any support for C++20.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
include bit unconditionally

It's a standard C++20 header.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: use contains

Found with readability-container-contains

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: remove pointless c_str

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: extra nodiscard

Found with modernize-use-nodiscard

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
clang-tidy: add missing static

Found with misc-use-internal-linkage

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use std::endian

Supported since GCC8.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use back()

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
use erase_if

Cast isspace to unsigned char for proper C API.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb committed:
remove filesystem linking

GCC and Clang greater than 8 don't need this. Remove it.

Signed-off-by: Rosen Penev <rosenp@gmail.com>

neheb commented:
I broke Cygwin somehow...

neheb commented:
```
--- a/src/xmp.cpp
+++ b/src/xmp.cpp
@@ -577,7 +577,7 @@ static XMP_Status nsDumper(void* refCon, XMP_StringPtr buffer, XMP_StringLen buf
   std::string out(buffer, bufferSize);
 
   // remove blanks: http://stackoverflow.com/questions/83439/remove-spaces-from-stdstring-in-c
-  out.erase(std::remove_if(out.begin(), out.end(), isspace), out.end());
+  std::erase_if(out, [](unsigned char c) { return std::isspace(c); });
 
   bool bURI = Internal::contains(out, "http://");
   bool bNS = Internal::contains(out, ':') && !bURI;
-- 

```

breaks cygwin. makes sense.

neheb committed:
use maybe_unused

Signed-off-by: Rosen Penev <rosenp@gmail.com>
